subsys/net/ip/net_shell.c

```diff
@@ -3563,6 +3566,51 @@ static int cmd_net_iface(const struct shell *shell, size_t argc, char *argv[])
 	return 0;
 }
 
+static int cmd_net_set_mac(const struct shell *sh, size_t argc, char *argv[])
+{
+#if !defined(CONFIG_NET_L2_ETHERNET)
+	PR_WARNING("MAC address can be set only for Ethernet\n");
+	return -ENOEXEC;
+#else
+	struct net_if *iface;
+	struct net_eth_addr mac_addr;
+	int idx;
+
+	if (argc < 3) {
+		PR_WARNING("Missing interface index and/or MAC address\n");
+	}
+
+	idx = get_iface_idx(sh, argv[1]);
+	if (idx < 0) {
+		return -ENOEXEC;
+	}
+
+	iface = net_if_get_by_index(idx);
+	if (!iface) {
+		PR_WARNING("No such interface in index %d\n", idx);
+		return -ENOEXEC;
+	}
+
+	if (net_if_l2(iface) != &NET_L2_GET_NAME(ETHERNET)) {
+		PR_WARNING("MAC address can be set only for Ethernet\n");
+		return -ENOEXEC;
+	}
+
+	if (net_bytes_from_str(mac_addr.addr, sizeof(mac_addr), argv[2]) < 0) {
+		PR_WARNING("Invalid MAC address\n");
+		return -ENOEXEC;
+	}
+
+	ret = net_if_set_link_addr(iface, mac_addr.addr, sizeof(mac_addr), NET_LINK_ETHERNET);
```

Comment:
The system will disallow changing of the MAC address if the interface is running. Perhaps we could check the error code here (-EPERM) and print a better message if that happens.